### ISSUE TITLE
`SHA-1 for file ... is not computed` when using `resolver.resolveRequest`

### ISSUE THREAD

**Issue by vjpr** [react/metro#330]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

I am using a custom `resolver.resolveRequest ` function in `rn-cli.config.js`. See https://facebook.github.io/metro/docs/en/configuration#resolver-options.

The error message is being printed from `metro/src/node-haste/DependencyGraph.js#getSha1`, which is being called by `metro/src/Bundler.js#transformFile`:

_metro/src/node-haste/DependencyGraph.js#getSha1_

```
  getSha1(filename: string): string {
    const resolvedPath = fs.realpathSync(filename);
    const sha1 = this._hasteFS.getSha1(resolvedPath);
    if (!sha1) {
      throw new ReferenceError(`SHA-1 for file ${filename} is not computed`);
    }
    return sha1;
  }
```

The issue is that the file won't exist in the haste map if you are resolving it in the `resolveRequest` hook.

**If the current behavior is a bug, please provide the steps to reproduce and a minimal repository on GitHub that we can `yarn install` and `yarn test`.**

https://github.com/vjpr/expo-v2-test-pnpm/tree/broken

You need to run `pnpm install` (the reason for the custom resolver is because symlinks are not supported by metro and hence cause problems for pnpm).

**What is the expected behavior?**

`getSha1` should compute hash if not in haste map.

**Please provide your exact Metro configuration and mention your Metro, node, yarn/npm version and operating system.**

metro@0.45.6
node@8.9.0
pnpm@2.21.1

## Workaround

Add to _metro/src/node-haste/DependencyGraph.js#getSha1_:

```
    if (!sha1) {
      return getFileHash(resolvedPath)
      function getFileHash(file) {
        return require('crypto')
          .createHash('sha1')
          .update(fs.readFileSync(file))
          .digest('hex')
      }
    }
```

I have released a patched version here: Use `metro-pnpm@0.45.6-vjpr.2`

**Comment by Stephen2** [comment]
I got this when trying to get Expo to work with yarn workspaces, using `expo-yarn-workspaces`

**Comment by pavermakov** [comment]
I've got the same issue when I tried to upgrade react-native to 0.57.0. I followed the steps mentioned the change log (https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md#057), but had no luck.

**Comment by mareksuscak** [comment]
@Stephen2 me too, did you solve it?

**Comment by Stephen2** [comment]
Sorry, moved to different project

**Comment by ekegodigital** [comment]
This is still a problem today with "react-native": "^0.59.2". Will there be real fix or to we need to copy/paste the workaround? 

+ BUNDLE_FILE=/Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app/main.jsbundle
+ node /Users/.../git/...-app-react-native/.../node_modules/react-native/cli.js bundle --entry-file index.js --platform ios --dev false --reset-cache --bundle-output /Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app/main.jsbundle --assets-dest /Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app
warning: the transform cache was reset.
Loading dependency graph, done.
error SHA-1 for file /Users/.../git/...-app-react-native/.../node_modules/metro/src/lib/polyfills/require.js (/Users/.../git/...-app-react-native/.../node_modules/metro/src/lib/polyfills/require.js) is not computed
+ [[ false != true ]]
+ [[ ! -f /Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app/main.jsbundle ]]
+ echo 'error: File /Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app/main.jsbundle does not exist. This must be a bug with'
error: File /Users/.../Library/Developer/Xcode/DerivedData/...-aexneivqfarcnddyzdbcbwoienmj/Build/Products/Release-iphoneos/....app/main.jsbundle does not exist. This must be a bug with
+ echo 'React Native, please report it here: https://github.com/facebook/react-native/issues'
React Native, please report it here: https://github.com/facebook/react-native/issues
+ exit 2

**Comment by nick-michael** [comment]
For anyone seeing this error when using `expo-yarn-workspaces`, I got this and fixed it by simply deleting the auto-generated file (the file that the error points to) and deleting all the node modules and then just running `yarn` again to re-install node modules and re-generate the file.

**Comment by esutton** [comment]
@vjpr How can I use your patched version in a react native project since metro is not included in my project's package.json ?

- I am not using expo.  
- I am seeing this after upgrading my react native project from 0.59 to 0.60.4.  

````
macOS 10.14.6
"react": "^16.9.0",
"react-native": "^0.60.4",

npm list -g --depth 0
/usr/local/lib
├── node@10.15.1
├── npm@6.10.3
└── react-native-cli@2.0.1
````

**Comment by dusan-dragon** [comment]
I am experiencing this on RN 0.60.5

**Comment by dioptre** [comment]
Also fixes RN 0.61

**Comment by dusan-dragon** [comment]
> Also fixes RN 0.61

Do you mean that on RN 0.61 it is working as expected?

**Comment by donholly** [comment]
Seeing the same issue as described on `0.60.5`. The workaround provided by @vjpr works - thank you! Looks like #455 should fix this. I'm curious why doesn't seem to be a bigger problem for everyone 😄

**Comment by dioptre** [comment]
@dusan-dragon I needed the @vjpr  fix to get RN 0.61 to work.

**Comment by ricick** [comment]
I'm having this issue on a windows 10 machine here. Same project on another dev machine with the same version of node doesn't display the issue.

**Comment by ikbal99** [comment]
put all files inside assets name of folder something like this

../../assets/beach.jpg

**Comment by vj-raghavan** [comment]
@donholly I am not sure why no one is flagging this issue more. I have just run in to this issue after I have setup a new terminal and cleaned up all my old node modules. 

Thanks to @vjpr workaround it worked after rebuilding the app. Is this a miss on our part ? or is this being actively looked into ?

**Comment by mhers91** [comment]
> put all files inside assets name of folder something like this
> 
> ../../assets/beach.jpg



> put all files inside assets name of folder something like this
> 
> ../../assets/beach.jpg

This worked for me ikbal99

**Comment by Gerlison** [comment]
Works for me, `0.61.5`

**Comment by kitttn** [comment]
0.61.5 indeed works. 
I also did `npm i -g react-native-cli` - it updated cli to 2.0.1 version and bundling worked.

**Comment by bkhurjawal** [comment]
I solved this issue by using `yarn start `instead of starting project with` react-native start`(it uses npm start). 
I have been adding dependencies in project using `Yarn add {dependencyName}` and starting project via npm. which created error of SHA-1.

**Comment by ricick** [comment]
I found it was an issue with me accessing the project on a second drive via a symlink on C: rather than navigating to the D: drive

**Comment by f4th4n** [comment]
I fixed it by install correct `react-native-cli`:

```
yarn global remove react-native
yarn global add react-native-cli
```

**Comment by tinmar33** [comment]
Thanks @f4th4n it works for me

**Comment by finghtingjie** [comment]
I had tried all the methods above but still not work for me.
OS:mac Catalina
node:13.3.0
watch:4.9.0
react-native:0.61.5

the error log like this

`Metro Bundler has encountered an error: SHA-1 for file /Users/.../Downloads/AwesomeProject/node_modules/metro/src/lib/polyfills/require.js (/Users/.../Downloads/AwesomeProject/node_modules/metro/src/lib/polyfills/require.js) is not computed`

**Comment by mohamedHamedDataCom** [comment]
i had the same issue and fix it by following this steps
1- updating  metro-react-native-babel-preset  to version 0.57.0  
 2- resetCache: true in metro.config 
3- removeing node modules & yarn lock file and reinstall

**Comment by finghtingjie** [comment]
> i had the same issue and fix it by following this steps
> 1- updating metro-react-native-babel-preset to version 0.57.0
> 2- resetCache: true in metro.config
> 3- removeing node modules & yarn lock file and reinstall

not work for me

**Comment by alancwoo** [comment]
I just ran into this issue while enjoying the fun of upgrading from `0.59.1` to `0.61.5` with `react-native-cli`.

My error made me realize I was running the bundler via a copy of `react-native` that was installed to `nvm` (Node Version Manager). Running `which react-native` confirmed this: `/Users/XXXXXX/.nvm/versions/node/v10.16.2/bin/react-native`

I made sure to remove all 'global' installs of react-native:

1. npm -g remove react-native
2. yarn global remove react-native
3. brew uninstall react-native-cli

Until `which react-native` responded with `react-native not found`

Then in my project, `npx react-native run-ios` worked.

**Comment by tourze** [comment]
same problem

**Comment by ymotse** [comment]
> I fixed it by install correct `react-native-cli`:
> 
> ```
> yarn global remove react-native
> yarn global add react-native-cli
> ```

Works for me!!! :D

**Comment by yashvanzara** [comment]
> I fixed it by install correct `react-native-cli`:
> 
> ```
> yarn global remove react-native
> yarn global add react-native-cli
> ```

`npm start` worked for me

**Comment by alincbiz** [comment]
I made an update to jdk, version 11, and it works for me

**Comment by acmedent** [comment]
I have installed again the react-native-cli and worked!

**Comment by smondal229** [comment]
> 0.61.5 indeed works.
> I also did `npm i -g react-native-cli` - it updated cli to 2.0.1 version and bundling worked.

currently my project is on 0.61.5 but it showing the same error

**Comment by leocavalcante** [comment]
Just got this with `"react-native": "0.61.5"`, I installed `react-native-cli` and `init`ialized a new project today 😛

**Comment by sayzenberg** [comment]
+1, ran into this with react-native 0.61.5 and react-native-cli 2.0.1. The only thing that's worked for me is the workaround provided by @vjpr.

**Comment by jhalvorson** [comment]
I've just ran into this with `react-native: 0.61.2`. @vjpr's workaround has temporarily fixed it for me.

I've tried lots of things including:
- deleting derived data
- removing react-native-cli's global install
- removing expo's global install (just in case)
- deleted the haste and metro temp directories
- restarted my system
- deleting node_modules and reinstalling

However, the above workaround is the only thing that's fixed it for me

**Comment by velsa** [comment]
Ran into the same issue. 

In my case, it complained about not being able to create SHA-1 for `timezones.json`, which is a package in my `node_modules` (https://www.npmjs.com/package/timezones.json). Since this package was a dependency of another package, my solution was to rename the directory `timezones.json` to `timezones-json` and also change the `require` in the corresponding package.

It seems that metro looks at the extension of the file and not at the file type, which caused it to think that a directory was a `.json` file.

**Comment by a7urag** [comment]
@vjpr solution worked for me, but any permanent solution for this?

**Comment by codejamninja** [comment]
@vjpr solution worked for me also. Is there a reason that code is not merged in?

**Comment by leandrosouzaa** [comment]
> I solved this issue by using `yarn start `instead of starting project with` react-native start`(it uses npm start).
> I have been adding dependencies in project using `Yarn add {dependencyName}` and starting project via npm. which created error of SHA-1.

@bkhurjawal  worked for me, thanks!!

**Comment by codejamninja** [comment]
I wrote a patch for it that works with the existing metro package.

1. Install the following dependency

    ```sh
    npm install --save mod-inline
    ```

2. Create the following file

    _prepare.js_
    ```js
    const fs = require('fs');
    const modInline = require('mod-inline').default;
    const path = require('path');

    const GET_FILE_HASH = / {4}if \(!sha1\) {\n {6}return getFileHash\(resolvedPath\);/;
    const GET_SHA1_FUNC = / {2}getSha1\(filename\) {((\n {4}.*|\n)+)+ {2}}/;
    const IF_SHA1 = / {4}if \(!sha1\) {/;

    const dependencyGraphPath = path.resolve(
      __dirname,
      '../node_modules/metro/src/node-haste/DependencyGraph.js'
    );
    const content = fs.readFileSync(dependencyGraphPath).toString();
    const patched = !!modInline.isolate(
      modInline.isolate(content, GET_SHA1_FUNC),
      GET_FILE_HASH
    );
    if (!patched) {
      const patchedContent = modInline.append(
        content,
        [GET_SHA1_FUNC, IF_SHA1],
        `
          return getFileHash(resolvedPath);
          function getFileHash(file) {
            return require('crypto')
              .createHash('sha1')
              .update(fs.readFileSync(file))
              .digest('hex');
          }`
      );
      fs.writeFileSync(dependencyGraphPath, patchedContent);
    }
    ```

3. Add the following script to the package.json

```json
{
  "scripts": {
    "prepare": "node ./prepare.js"
  }
}
```

**Comment by alexyazvinsky** [comment]
same issue, tried on several osx
and only @vjpr's solution helped
any other workarounds?
and why is it not fixed officially?

OS:mac Catalina 10.15.4
node:10.16.3
watchman:4.9.0
react-native:0.62.2

**Comment by a7urag** [comment]
Finally got what went wrong here, folder name has 1 upper case character in require. eg. `require(../../../assets/marketPlace/banner.png)`. Here "p" was upper case here but folder was with lower case p. After changing require path it was working again.

**Comment by coolcorexix** [comment]
Hello, I have this problem and then fixed it by using directly the 
`node_modules/react-native/scripts/launchPackager.command` instead of `react-native start`

If you have any further investigation on this please let me know

**Comment by SeanHayes** [comment]
I have this same error with the following code:

```
import { createIconSet } from '@expo/vector-icons';

const glyphMap = {
    'kk-logo': 0xf101,
};
const expoAssetId = require("../.cache/iconfont/iconfont.ttf");
const CustomIcon = createIconSet(glyphMap, 'iconfont', expoAssetId);

export default CustomIcon;
```

`../.cache/iconfont/iconfont.ttf` is a custom font generated using svgs.

It was working fine, then my Macbook restarted after an update, now I get this error no matter what I do. Tried clearing caches, tmp directories, etc. but nothing.

**Comment by samuelneff** [comment]
I created a new blank React Native app and got this same error about SHA1 and `require.js`. My React Native application name had a mixed-case name, ex. `MyApp`. Recreated the same app a half dozen times with different versions of node and React Native and same error each time.

Then I created the same app but named it `myapp` and didn't have this problem. It just worked.

**Comment by dannycoulombe** [comment]
Thanks @vjpr, you saved me hours. I can't believe Github doesn't offer some sort of donation button.

**Comment by vjpr** [comment]
Just an update on why this happens.

Here is the function that is generating the message:

https://github.com/facebook/metro/blob/master/packages/metro/src/node-haste/DependencyGraph.js#L208

```js
  getSha1(filename) {
    // TODO If it looks like we're trying to get the sha1 from a file located
    // within a Zip archive, then we instead compute the sha1 for what looks
    // like the Zip archive itself.
    const splitIndex = filename.indexOf(".zip/");
    const containerName =
      splitIndex !== -1 ? filename.slice(0, splitIndex + 4) : filename; // TODO Calling realpath allows us to get a hash for a given path even when
    // it's a symlink to a file, which prevents Metro from crashing in such a
    // case. However, it doesn't allow Metro to track changes to the target file
    // of the symlink. We should fix this by implementing a symlink map into
    // Metro (or maybe by implementing those "extra transformation sources" we've
    // been talking about for stuff like CSS or WASM).

    const resolvedPath = fs.realpathSync(containerName);

    const sha1 = this._hasteFS.getSha1(resolvedPath);

    if (!sha1) {
      throw new ReferenceError(
        `SHA-1 for file ${filename} (${resolvedPath}) is not computed`
      );
    }

    return sha1;
  }

```

_jest-haste-map/build/worker.js_

```
function _getSha() {
  _getSha = _asyncToGenerator(function*(data) {
    const sha1 = data.computeSha1
      ? sha1hex(_gracefulFs().default.readFileSync(data.filePath))
      : null;
    return {
      dependencies: undefined,
      id: undefined,
      module: undefined,
      sha1
    };
  });
  return _getSha.apply(this, arguments);
}
```

When `metro` resolves a file it looks for the file in a `jest-haste-map` which also watches for changes to these files. There are two implementations for building a haste map (crawling the file tree from the project roots):

1. Facebook's `watchman` native (c++, rust, etc.) library. 
   _jest-haste-map/build/crawlers/watchman.js_
2. Node.js implementation
    _jest-haste-map/build/crawlers/node.js_

Both don't support _following_ symlinks...so if you have any symlinks in your project root that point to files outside of the project root, then they will never make it into the haste map which also means they will never be watched. It was the first feature request for `metro` 5 years ago and its unresolved. You can see the issue here: https://github.com/facebook/metro/issues/1#issuecomment-641633646

So there are two situations here.

a. Your symlink points to a file that is inside your project root.
b. Your symlink points to a file outside of your project root.
     E.g. If you are using a pnpm-based monorepo, all your symlinks resolve to the monorepo root. So if `expo` is nested inside the monorepo, then you symlinks will always point outside the project root.

For `a`, your `sha-1` will actually be found, because your file's realpath will be in the haste map.
For `b`, your `sha-1` will not be found because your file's realpath is not in the haste map.

```js
const resolvedPath = fs.realpathSync(containerName);
```

When `metro` requires a file it needs to transform it. It first checks if its transformation is cached in a `metro-cache`. The `sha-1` of the file is used as part of the cache key.

If its not in the cache, the file contents is actually read not from the haste map but from the file system. So if your file was symlinked to outside of a project root, it won't be in the haste map at all, and the sha1 can't be accessed...but it will still be able to be loaded fine.

https://github.com/facebook/metro/blob/7814c2840c49c16788041284fd65df25aa997d8c/packages/metro/src/DeltaBundler/Transformer.js#L141-L147

**Comment by vjpr** [comment]
I think I found a simple explanation for this problem. See https://github.com/facebook/jest/issues/10063#issuecomment-642356808.

When `metro` talks to `watchman` it sends a `query` with a `glob` option. When a `glob` is used it prevents files and folders starting with a dot from being returned by the watchman query. There is an option to pass called `glob_includedotfiles` which includes dot files.

Depending on whether the clock is available, if `glob` is not used, then you are not going to get any dotfiles or dot dirs included.

So this explains the sporadic behaviour.

```js
const relativeRoot = fastPath.relative(rootDir, root);
const query = clocks.has(relativeRoot) // Use the `since` generator if we have a clock available
  ? {
      expression,
      fields,
      since: clocks.get(relativeRoot)
    } // Otherwise use the `glob` filter
  : {
      expression,
      fields,
      glob
    };
const response = yield cmd('query', root, query);
```

This explains why @SeanHayes [was having a problem](https://github.com/facebook/metro/issues/330#issuecomment-626232472) until his MacBook restarted.

And it also explains why `pnpm` doesn't work because all its deps are found in `node_modules/.pnpm` which includes a dir starting with a dot.

Its a one line fix that [is in the pipeline](https://github.com/facebook/jest/issues/10063#issuecomment-633026769) but I feel like I went to the ends of the earth to track it down.

**Comment by wannymiarelli** [comment]
@vjpr 

You are right about symlinks, on windows I had this problem with the "Documents" directory. It is spelt "Documenti" in Italian if you run the "yarn start" using the "Documenti" alias instead of the "Documents" it fails with the SHA-1 computed error.

So c:/users/[username]/Documenti - (Italian version) fails
c:/users/[username]/Documents - works

Running the command from the "Documents" path works just fine.

**Comment by jinyuyoulong** [comment]
I make it  work for this:
` npx react-native bundle --platform ...`
but not work when I use like this:
`react-native bundle --platform`

didn't know why!

**Comment by Pat-Gekoski** [comment]
@jinyuyoulong I did the same thing but I prepended it with yarn instead of npx. Thank you for your advice
I used 
yarn react-native bundle --platform android --dev false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res/

**Comment by leandrosouzaa** [comment]
Try ```yarn start``` instead of ```yarn react-native....```

**Comment by Nargle30** [comment]
_npm cache clean --force
cd ios/  && rm -rf build/ && xcodebuild clean
npm start -- --reset-cache_

Worked for me

**Comment by saSinclair** [comment]
Had the same issue after doing some folder creations and files relocation - moving an image to a new directory assets/images/image.png. I would get error 

SHA-1 for file <IMAGE LOCATION PATH> is not computed

After moving the image to assets/image.png - it seems to go away O_0

**Comment by umar-muneer** [comment]
> I solved this issue by using `yarn start `instead of starting project with` react-native start`(it uses npm start).
> I have been adding dependencies in project using `Yarn add {dependencyName}` and starting project via npm. which created error of SHA-1.

this did it for me as well. i was running project directly `react-native start` just did `yarn start`

**Comment by ethans333** [comment]
The solution for me was as simple as moving my images file into the same directory as the ```App.js``` script.

**Comment by devheedoo** [comment]
It happened when:

`$ react-native start` → connecting with device → just before finishing bundling

**f4th4n**'s solution worked to me (with npm):

```
$ npm install -g --force react-native-cli
```

**Comment by changkem** [comment]
it happened when: 

- metro not configure extraNodeModules as same as follows code
```
extraNodeModules: {
     'react-native': your ReactNative path,
 };
```
- react native configure reactNativePath

```
reactNativePath: your ReactNative path
```

I fixed this problem by remove extraNodeModules and reactNativePath or unified them

**Comment by isameera** [comment]
just use npx
Ex: npx react-native bundle --dev false --platform android --entry-file index.js --bundle-output ./android/app/src/main/assets/index.android.bundle --assets-dest ./android/app/src/main/res

**Comment by DouglasAzevedo** [comment]
A solução proposta por @vjpr funcionou perfeitamente pra mim.

**Comment by rpcarnell** [comment]
The solution offered by @vjpr worked like a charm.

**Comment by khomchyk** [comment]
Hello!
Are there any plans to fix this issue or support searching for symlinks?

The patch suggested by @vjpr is working, but it will be better to have an official support of this functionality.

**Comment by agelessZeal** [comment]
npm i -g react-native-cli
It is solution of this issue.

**Comment by rad182** [comment]
If anyone else having issue when archiving ios.
I had to install react-native-cli and use it.
`yarn add react-native-cli`
then override CLI_PATH
![image](https://user-images.githubusercontent.com/71201/107841142-6a219b80-6df3-11eb-9491-54b900816fab.png)

**Comment by huzaifaaak** [comment]
I am getting this issue on rn-tester app on android. None of the above is working for me

**Comment by JulesGuesnon** [comment]
Hi there !

My case is probably unlikely: the folder containing the project had `.zip` in the name and was causing the issue.

I checked the code and there's a specific case for `.zip` files in `getSha1` which caused the error

**Comment by vjpr** [comment]
If you are using pnpm, see https://github.com/pnpm/pnpm/issues/3731.

**Comment by zaidraddad94** [comment]
you ar the man

**Comment by muganwas** [comment]
> put all files inside assets name of folder something like this
> 
> ../../assets/beach.jpg

This worked for me, thanks!

**Comment by FoolCoder** [comment]
i am having this issue while generating debug apk! why?

**Comment by patricksantos-bcs** [comment]
> If anyone else having issue when archiving ios. I had to install react-native-cli and use it. `yarn add react-native-cli` then override CLI_PATH ![image](https://user-images.githubusercontent.com/71201/107841142-6a219b80-6df3-11eb-9491-54b900816fab.png)

this one works for me!

**Comment by patricksantos-bcs** [comment]
> If anyone else having issue when archiving ios. I had to install react-native-cli and use it. `yarn add react-native-cli` then override CLI_PATH ![image](https://user-images.githubusercontent.com/71201/107841142-6a219b80-6df3-11eb-9491-54b900816fab.png)

this one works for me!

> If anyone else having issue when archiving ios. I had to install react-native-cli and use it. `yarn add react-native-cli` then override CLI_PATH ![image](https://user-images.githubusercontent.com/71201/107841142-6a219b80-6df3-11eb-9491-54b900816fab.png)

This one works for me! Thanks!

**Comment by Vednus** [comment]
Just an FYI: was having this issue when running codepush and fixed it by restarting my machine.

**Comment by kouohhashi** [comment]
Hi, I have similar issue.

My react-native version is 0.67 and gave me similar `error SHA-1 for ...node_modules/metro-runtime/src/polyfills/require.js" not computed` error. I don't know what to try next...

npm react-native run-android works on my device.
I'm pretty sure i have not changes significant changes instead of key info...

What is the root cause and how i can find out fix?

Thanks in advance.

**Comment by instanceof-inf** [comment]
None of the above worked for me, the only thing that worked was checking the 'For install builds only' checkbox under 'Bundle React Native code and images' under Build Phases
<img width="574" alt="Screenshot 2022-11-22 at 22 21 36" src="https://user-images.githubusercontent.com/24798224/203413554-4d398d62-8a0b-42f7-b831-a425f75c343a.png">

**Comment by fwielstra** [comment]
Just an addition, I ran into this issue today; at first I thought it was due to having pages and routers of my app spread across different libraries using NX, which uses a symlink to `node_modules` in every app project.

However, the issue was that I already had Metro running in another terminal tab, and kept trying to remove `node_modules` and restart the app (and Metro) in another tab.

TL;DR make sure there are no other Metro processes running when you try anything described in this thread.

**Comment by dezoxel** [comment]
> I solved this issue by using `yarn start `instead of starting project with` react-native start`(it uses npm start). I have been adding dependencies in project using `Yarn add {dependencyName}` and starting project via npm. which created error of SHA-1.

I have the same issue.
React Native v0.71.2.

Running thru `yarn start` solves the issue, thanks a lot!

**Comment by lukasgjetting** [comment]
Started working again after I ran `watchman watch-del '{PROJECT_DIR}` where `{PROJECT_DIR}` is the root directory of your RN project, i.e. `/Users/lukas/projects/myapp`

**Comment by yoannes** [comment]
RN 0.72.3

I'm using yarn workspaces and the following worked for me:

  "workspaces": {
    "packages": [
      "packages/*"
    ],
    "nohoist": [
      "**/react-native",
      "**/react-native/**",
      "**/@react-native",
      "**/@react-native/**"
    ]
  },

**Comment by spsaucier** [comment]
I was getting this on an Expo project when the `start` command was running `react-native` instead of `expo` directly

**Comment by robhogan** [comment]
It looks like there are a few different things reported here, but focusing on the original issue - Metro's architecture requires all files to be visible to Metro by either being under [`projectRoot`](https://metrobundler.dev/docs/configuration/#projectroot) or one of your configured [`watchFolders`](https://metrobundler.dev/docs/configuration/#watchfolders). 

That's true even of files you resolve yourself with a custom `resolveRequest`, because it's how Metro knows when a file changes and when to invalidate transform caches. The error comes from that mechanism - if Metro isn't watching your file, it doesn't know the hash, and can't look it up in the transform cache.

That covers the original issue - the behaviour is as designed (though since this issue we've improved the error message).

If folks are hitting similar errors not explained by the above, please open a new issue.
